packages/nuxt/src/pages/runtime/router.options.ts

```diff
@@ -11,43 +11,70 @@ type ScrollPosition = Awaited<ReturnType<RouterScrollBehavior>>
 
 // Default router options
 // https://router.vuejs.org/api/#routeroptions
-export default <RouterConfig> {
-  scrollBehavior (to, from, savedPosition) {
+export default <RouterConfig>{
+  scrollBehavior: (
+    to: RouteLocationNormalized,
+    from: RouteLocationNormalized,
+    savedPosition: ScrollPosition | null
+  ) => {
+    // Check if the navigation is the first page load
+    const isFirstLoad = from === START_LOCATION
     const nuxtApp = useNuxtApp()
     // @ts-expect-error untyped, nuxt-injected option
-    const behavior = useRouter().options?.scrollBehaviorType ?? 'auto'
+    const scrollBehaviourType = useRouter().options?.scrollBehaviorType ?? 'auto'
 
-    // By default when the returned position is falsy or an empty object, vue-router will retain the current scroll position
-    // savedPosition is only available for popstate navigations (back button)
-    let position: ScrollPosition = savedPosition || undefined
+    // Handle page reload scenario where the saved position should be used
+    if (isFirstLoad && savedPosition) {
+      // Restore browser's default scroll behavior to auto (vue-router sets this when using custom ScrollBehavior)
+      // to avoid page jumps on load
+      window.history.scrollRestoration = 'auto'
+      return savedPosition
+    } else {
+      // For other navigations, ensure manual scrollRestoration
+      window.history.scrollRestoration = 'manual'
+    }
 
+    // Check if the route explicitly disables automatic scroll to top
     const routeAllowsScrollToTop = typeof to.meta.scrollToTop === 'function' ? to.meta.scrollToTop(to, from) : to.meta.scrollToTop
-
-    // Scroll to top if route is changed by default
-    if (!position && from && to && routeAllowsScrollToTop !== false && isChangingPage(to, from)) {
-      position = { left: 0, top: 0 }
+    if (routeAllowsScrollToTop === false) {
+      return false // Do not scroll to top if the route disallows it
     }
 
-    // Hash routes on the same page, no page hook is fired so resolve here
-    if (to.path === from.path) {
-      if (from.hash && !to.hash) {
-        return { left: 0, top: 0 }
+    const calculatePosition = (behavior: ScrollBehavior): ScrollPosition => {
```

Comment:
Could we move this out of the function and pass the `to` as well? 🤔

Reply:
Also needs `savedPosition`.

I've moved it out of the function now. I agree with you for the sake of code readability.